src/engine/main.cpp

```diff
@@ -76,18 +159,28 @@ int main( int argc, char **argv )
             return 0;
         };
 
+        InputQuery inputQuery = prepareInputQuery();
```

Comment:
To avoid unnecessary calls of the copy constructor, maybe declare the prepareInputQuery method like this:

`void prepareInputQury(InputQuery &inputQuery);
`
The method will update the given reference of the inputQuery object, which is created in the main function, like this:

```
InputQuery inputQuery; 
prepareInputQuery(inputQuery);

```

Reply:
Hmm so I've tried to do this, but it's pretty tricky. In particularly we eventually end up having to push that change all the way through to the method `QueryLoader.loadQuery`. From there, that method then gets called via the Python/C interface which I have no idea how to update properly.

I've added a comment to the effect that this change _should_ be done. Is there anything else I can do to improve this without meddling with the Python/C interface?

Reply:
Hum.. Ideally this change can be pushed through because we don't want to make unnecessary copies. 
This is not an issue for small networks, but for larger network where one only wants to do bound propagation on this without search, copying the input query can become a runtime and memory bottleneck.